Re-renders of memoized components on shallow route changes

## Describe the bug
As we discussed in #983, components that utilise useTranslation are always re-rendered on shallow route changes, regardless of if they are memoized or not. This is caused by the fact that the i18n instance is re-created every time the App is rendered.

The problem has been exacerbated in the last two weeks: it now re-renders the memoized component **twice** on every shallow route change rather than once (compare 8.0.1 and 8.1.1). I suspect this is because of this MR being merged in react-i18next: https://github.com/i18next/react-i18next/pull/1273 - which is also mentioned in  #1023.
## Occurs in next-i18next version
8.1.1

## Steps to reproduce
Example repo: https://github.com/skrivanos/next-i18next-rerender-poc
https://next-i18next-rerender-poc.vercel.app

## Expected behaviour
The memoized component should not re-render.

@isaachinman I wanted to work on this myself, but ran into some issues with `yarn link`... If you could point me in the right direction (since Google isn't helping) I'd happily contribute. After running `yarn link next-i18next` which correctly symlinks the package to my local copy, I get these errors:

```
info  - ready on http://localhost:3000
event - build page: /
wait  - compiling...
warn  - ../Code/github.com/skrivanos/next-i18next/node_modules/i18next-fs-backend/cjs/readFile.js
Critical dependency: the request of a dependency is an expression

../Code/github.com/skrivanos/next-i18next/node_modules/i18next-fs-backend/cjs/readFile.js
Critical dependency: the request of a dependency is an expression

../Code/github.com/skrivanos/next-i18next/dist/commonjs/serverSideTranslations.js
Critical dependency: the request of a dependency is an expression
info  - ready on http://localhost:3000
Error: Cannot find module '/Users/alex/my-app/next-i18next.config.js'
    at webpackEmptyContext (eval at ../Code/github.com/skrivanos/next-i18next/dist/commonjs sync recursive (/Users/alex/my-app/.next/server/pages/index.js:103:1), <anonymous>:2:10)
    at eval (webpack-internal:///../Code/github.com/skrivanos/next-i18next/dist/commonjs/serverSideTranslations.js:90:156)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  code: 'MODULE_NOT_FOUND'
}
```

--- a/src/appWithTranslation.client.test.tsx
+++ b/src/appWithTranslation.client.test.tsx
@@ -2,8 +2,10 @@ import React from 'react'
 import fs from 'fs'
 import { screen, render } from '@testing-library/react'
 import { I18nextProvider } from 'react-i18next'
+import createClient from './createClient'
 
 import { appWithTranslation } from './appWithTranslation'
+import { NextRouter } from 'next/router'
 
 jest.mock('fs', () => ({
   existsSync: jest.fn(),
@@ -19,15 +21,16 @@ jest.mock('react-i18next', () => ({
   __esmodule: true,
 }))
 
+jest.mock('./createClient', () => jest.fn())
 
 const DummyApp = appWithTranslation(() => (
   <div>Hello world</div>
 ))
 
-const props = {
+const createProps = (locale = 'en', router: Partial<NextRouter> = {}) => ({
   pageProps: {
     _nextI18Next: {
-      initialLocale: 'en',
+      initialLocale: locale,
       userConfig: {
         i18n: {
           defaultLocale: 'en',
@@ -36,9 +39,15 @@ const props = {
       },
     },
   } as any,
-} as any
+  router: {
+    locale: locale,
+    route: '/',
+    ...router,
+  },
+} as any)
 
-const renderComponent = () =>
+const defaultRenderProps = createProps()
+const renderComponent = (props = defaultRenderProps) =>
   render(
     <DummyApp
       {...props}
@@ -50,6 +59,8 @@ describe('appWithTranslation', () => {
     (fs.existsSync as jest.Mock).mockReturnValue(true);
     (fs.readdirSync as jest.Mock).mockReturnValue([]);
     (I18nextProvider as jest.Mock).mockImplementation(DummyI18nextProvider)
+    const actualCreateClient = jest.requireActual('./createClient');
+    (createClient as jest.Mock).mockImplementation(actualCreateClient)
   })
   afterEach(jest.resetAllMocks)
 
@@ -69,6 +80,7 @@ describe('appWithTranslation', () => {
       },
     } as any)
     const customProps = {
+      ...createProps(),
       pageProps: {
         _nextI18Next: {
           initialLocale: 'en',
@@ -91,6 +103,7 @@ describe('appWithTranslation', () => {
       <div>Hello world</div>
     ))
     const customProps = {
+      ...createProps(),
       pageProps: {
         _nextI18Next: {
           initialLocale: 'en',
@@ -122,6 +135,37 @@ describe('appWithTranslation', () => {
 
     expect(fs.existsSync).toHaveBeenCalledTimes(0)
     expect(fs.readdirSync).toHaveBeenCalledTimes(0)
+  })
+
+  it('should use locale from router', () => {
+    renderComponent(createProps('de'))
+    const [args] = (I18nextProvider as jest.Mock).mock.calls
+    expect(args[0].i18n.language).toEqual('de')
+  })
+
+  it('does not re-call createClient on re-renders unless locale or props have changed', () => {
+    const { rerender } = renderComponent()
+    expect(createClient).toHaveBeenCalledTimes(1)
+    rerender(
+      <DummyApp
+        {...defaultRenderProps}
+      />
+    )
+    expect(createClient).toHaveBeenCalledTimes(1)
+    const newProps = createProps()
+    rerender(
+      <DummyApp
+        {...newProps}
+      />
+    )
+    expect(createClient).toHaveBeenCalledTimes(2)
+    newProps.router.locale = 'de'
+    rerender(
+      <DummyApp
+        {...newProps}
+      />
+    )
+    expect(createClient).toHaveBeenCalledTimes(3)
   })
 
 })

--- a/src/appWithTranslation.server.test.tsx
+++ b/src/appWithTranslation.server.test.tsx
@@ -40,6 +40,10 @@ const props = {
       },
     },
   } as any,
+  router: {
+    locale: 'en',
+    route: '/',
+  },
 } as any
 
 const renderComponent = () =>

--- a/src/appWithTranslation.tsx
+++ b/src/appWithTranslation.tsx
@@ -46,7 +46,8 @@ export const appWithTranslation = (
         throw new Error('appWithTranslation was called without config.i18n')
       }
 
-      return createClient({
+
+      const instance = createClient({
         ...createConfig({
           ...userConfig,
           lng: locale,
@@ -54,11 +55,11 @@ export const appWithTranslation = (
         lng: locale,
         resources: initialI18nStore,
       }).i18n
-    }, [_nextI18Next, locale])
 
-    useMemo(() => {
-      globalI18n = i18n
-    }, [i18n])
+      globalI18n = instance
+
+      return instance
+    }, [_nextI18Next, locale])
 
     return i18n !== null ? (
       <I18nextProvider

--- a/src/appWithTranslation.tsx
+++ b/src/appWithTranslation.tsx
@@ -22,12 +22,17 @@ export const appWithTranslation = (
   configOverride: UserConfig | null = null,
 ) => {
   const AppWithTranslation = (props: AppProps) => {
-    let i18n: I18NextClient | null = null
-    let locale = null
+    const { _nextI18Next } = props.pageProps
+    const { locale } = props.router
 
-    if (props?.pageProps?._nextI18Next) {
-      let { userConfig } = props.pageProps._nextI18Next
-      const { initialI18nStore, initialLocale } = props.pageProps._nextI18Next
+    // Memoize the instance and only re-initialize when either:
+    // 1. The route changes (non-shallowly)
+    // 2. Router locale changes
+    const i18n: I18NextClient | null = useMemo(() => {
+      if (!locale || !_nextI18Next) return null
+
+      let { userConfig } = _nextI18Next
+      const { initialI18nStore } = _nextI18Next
 
       if (userConfig === null && configOverride === null) {
         throw new Error('appWithTranslation was called without a next-i18next config')
@@ -41,17 +46,15 @@ export const appWithTranslation = (
         throw new Error('appWithTranslation was called without config.i18n')
       }
 
-      locale = initialLocale;
-
-      ({ i18n } = createClient({
+      return createClient({
         ...createConfig({
           ...userConfig,
-          lng: initialLocale,
+          lng: locale,
         }),
-        lng: initialLocale,
+        lng: locale,
         resources: initialI18nStore,
-      }))
-    }
+      }).i18n
+    }, [_nextI18Next, locale])
 
     useMemo(() => {
       globalI18n = i18n

--- a/src/serverSideTranslations.test.tsx
+++ b/src/serverSideTranslations.test.tsx
@@ -18,7 +18,6 @@ const DummyApp = appWithTranslation(() => (
 const props = {
   pageProps: {
     _nextI18Next: {
-      initialLocale: 'en',
       userConfig: {
         i18n: {
           defaultLocale: 'en',
@@ -185,7 +184,6 @@ describe('serverSideTranslations', () => {
         initialI18nStore: {
           'en-US': {},
         },
-        initialLocale: 'en-US',
         userConfig: {
           i18n: {
             defaultLocale: 'en-US',

--- a/src/serverSideTranslations.test.tsx
+++ b/src/serverSideTranslations.test.tsx
@@ -27,6 +27,9 @@ const props = {
       },
     },
   } as any,
+  router: {
+    locale: 'en',
+  },
 } as any
 
 const renderDummyComponent = () =>

--- a/src/serverSideTranslations.ts
+++ b/src/serverSideTranslations.ts
@@ -106,7 +106,6 @@ export const serverSideTranslations = async (
   return {
     _nextI18Next: {
       initialI18nStore,
-      initialLocale,
       userConfig: config.serializeConfig ? userConfig : null,
     },
   }

--- a/src/types.ts
+++ b/src/types.ts
@@ -52,7 +52,6 @@ export type CreateClientReturn = {
 export type SSRConfig = {
   _nextI18Next: {
     initialI18nStore: any
-    initialLocale: string
     userConfig: UserConfig | null
   }
 }

